### PR TITLE
Adding hydraulic fracture examples with poroelasticity

### PR DESCRIPTION
**Pull request by jhuang2601**
This PR adds

- [x] KGD frac toughness dominated case with poroelasticity
- [x] pennyShaped frac viscosity dominated case with poroelasticity
- [x] pennyShaped frac toughness dominated case with poroelasticity
- [x] PKN frac viscosity dominated case with poroelasticity
- [x] add smoke tests into integratedTests 
       https://github.com/GEOS-DEV/integratedTests/pull/19
- [x] fix convergence issue for running `pknViscosityDominated_benchmark.xml`

### CODE REVIEW

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp [hunk 1]
```diff
@@ -92,7 +92,7 @@ class Hydrofracture : public MGRStrategyBase< 1 >
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetPrintLevel( mgrData.coarseSolver.ptr, 0 ) );
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxIter( mgrData.coarseSolver.ptr, 1 ) );
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetTol( mgrData.coarseSolver.ptr, 0.0 ) );
-    GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxLevels( mgrData.coarseSolver.ptr, 25 ) );
+    GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxLevels( mgrData.coarseSolver.ptr, 1 ) );
```
Comment (castelletto1) [inline src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp:95]:
```suggestion
    GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMinCoarseSize( mgrData.coarseSolver.ptr, 1000 ) );
```

Could you try this options and check if you observe any improvement? If yes, we can further investigate whether `1000` is the appropriate minimum size. Thanks.

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp:95]: @castelletto1 I tested this option, unfortunately, convergence issue persists for `pknViscosityDominated_benchmark.xml`
Below is the error message from running log:
```
++ Fracture propagation. Re-entering Newton Solve.
    Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
    ( R ) = ( 7.88e+00 ) ; 
{ 3, 1 }
		Linear Solver | Success | Iterations: 40 | Final Rel Res: 0.000473497 | Make Restrictor Time: 0 | Compute Auu Time: 0 | SC Filter Time: 0 | Setup Time: 3.15068 s | Solve Time: 7.87251 s
    Attempt:  0, ConfigurationIter:  0, NewtonIter:  1
    ( R ) = ( 2.25e+12 ) ; 
    Last LinSolve(iter,res) = (  40, 4.73e-04 ) ; 
    The residual norm is above the maxAllowedResidualNorm of 1000000000. Newton loop terminated.
New dt = 5
    Attempt:  1, ConfigurationIter:  0, NewtonIter:  0
    ( R ) = ( 1.23e+09 ) ; 
    The residual norm is above the maxAllowedResidualNorm of 1000000000. Newton loop terminated.
New dt = 2.5
    Attempt:  2, ConfigurationIter:  0, NewtonIter:  0
    ( R ) = ( 6.13e+08 ) ; 
{ 3, 1 }
		Linear Solver | Success | Iterations: 26 | Final Rel Res: 0.000667621 | Make Restrictor Time: 0 | Compute Auu Time: 0 | SC Filter Time: 0 | Setup Time: 2.97576 s | Solve Time: 5.13145 s
    Attempt:  2, ConfigurationIter:  0, NewtonIter:  1
    ( R ) = ( 1.25e+04 ) ; 
    Last LinSolve(iter,res) = (  26, 6.68e-04 ) ; 
{ 3, 1 }
		Linear Solver | Success | Iterations: 47 | Final Rel Res: 0.000429517 | Make Restrictor Time: 0 | Compute Auu Time: 0 | SC Filter Time: 0 | Setup Time: 3.05405 s | Solve Time: 9.19932 s
    Attempt:  2, ConfigurationIter:  0, NewtonIter:  2
    ( R ) = ( 5.20e+03 ) ; 
    Last LinSolve(iter,res) = (  47, 4.30e-04 ) ; 
{ 3, 1 }
		Linear Solver | Success | Iterations: 54 | Final Rel Res: 0.000665758 | Make Restrictor Time: 0 | Compute Auu Time: 0 | SC Filter Time: 0 | Setup Time: 2.94388 s | Solve Time: 10.6859 s
    Attempt:  2, ConfigurationIter:  0, NewtonIter:  3
    ( R ) = ( 9.19e+13 ) ; 
    Last LinSolve(iter,res) = (  54, 6.66e-04 ) ; 
    The residual norm is above the maxAllowedResidualNorm of 1000000000. Newton loop terminated.
New dt = 1.25
    Attempt:  3, ConfigurationIter:  0, NewtonIter:  0
    ( R ) = ( 7.14e+10 ) ; 
    The residual norm is above the maxAllowedResidualNorm of 1000000000. Newton loop terminated.
New dt = 0.625
    Attempt:  4, ConfigurationIter:  0, NewtonIter:  0
    ( R ) = ( 3.57e+10 ) ; 
    The residual norm is above the maxAllowedResidualNorm of 1000000000. Newton loop terminated.
New dt = 0.3125
Convergence not achieved.
***** ERROR
***** LOCATION: /shared/scratch1/Users/j0551570/geosx/IntegratedTest/GEOS/src/coreComponents/physicsSolvers/SolverBase.cpp:763
***** Controlling expression (should be false): true
***** Rank 4: Nonconverged solutions not allowed. Terminating...
```

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp:95]: @castelletto1 @CusiniM Should `maxAllowedResidualNorm` be boosted for nonlinear solver? as `maxAllowedResidualNorm="1e+15"`?

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp]: Finally, increasing `maxAllowedResidualNorm` solves the problem

Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp]: Nice, thank you. What is linear solver option?

- `GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxLevels( mgrData.coarseSolver.ptr, 1 ) )`
- `GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMinCoarseSize( mgrData.coarseSolver.ptr, 1000 ) )`

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp]: @castelletto1 `GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMinCoarseSize( mgrData.coarseSolver.ptr, 1000 ) )` is used

Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp]: Out of curiosity, can you check what happens if you revert back to setting the max number of levels to `1`? Thank you

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/Hydrofracture.hpp]: @castelletto1 The PKN case can also be successfully completed with option 1 (`GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxLevels( mgrData.coarseSolver.ptr, 1 ) )`)

**Comment by herve-gross** [comment]
@Guotong-Ren can you take a look and try if the cases run on your side? Thank you very much.

**Comment by Guotong-Ren** [comment]
> @Guotong-Ren can you take a look and try if the cases run on your side? Thank you very much.

**Review by Guotong-Ren — APPROVED** [review]
I tested penny-shaped and KGD poroelastic FP. Results look good to me.

**Comment by jhuang2601** [comment]
> I tested penny-shaped and KGD poroelastic FP. Results look good to me.

Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?

**Comment by Guotong-Ren** [comment]
> > I tested penny-shaped and KGD poroelastic FP. Results look good to me.
> 
> Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?

I only see only one time step size cut for both viscosity and toughness dominated PKN cases. The NL solver performs well I tight the convergence criterion to 1e-6/-7 and each time step solve takes about 10 iterations to converge.

**Comment by jhuang2601** [comment]
> > > I tested penny-shaped and KGD poroelastic FP. Results look good to me.
> > 
> > 
> > Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?
> 
> I only see only one time step size cut for both viscosity and toughness dominated penny shape cases. The NL solver performs well I tight the convergence criterion to 1e-6/-7 and each time step solve takes about 10 iterations to converge. I did not run PKN cases yet. Do you see any performance issues there?

In this PR, I slightly updated the MGR recipe for HF; otherwise, un-convergence issue will be faced for PKN cases.

**Review by herve-gross — APPROVED** [review]

**Comment by Guotong-Ren** [comment]
> > > > I tested penny-shaped and KGD poroelastic FP. Results look good to me.
> > > 
> > > 
> > > Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?
> > 
> > 
> > I only see only one time step size cut for both viscosity and toughness dominated penny shape cases. The NL solver performs well I tight the convergence criterion to 1e-6/-7 and each time step solve takes about 10 iterations to converge. I did not run PKN cases yet. Do you see any performance issues there?
> 
> In this PR, I slightly updated the MGR recipe for HF; otherwise, un-convergence issue will be faced for PKN cases.

I also saw enormous residuals when debugging the quasi-newton for the penny shaped case, do you have any idea which equations are associated with this big number?

**Review by CusiniM — APPROVED** [review]

**Comment by Guotong-Ren** [comment]
> > > > > I tested penny-shaped and KGD poroelastic FP. Results look good to me.
> > > > 
> > > > 
> > > > Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?
> > > 
> > > 
> > > I only see only one time step size cut for both viscosity and toughness dominated penny shape cases. The NL solver performs well I tight the convergence criterion to 1e-6/-7 and each time step solve takes about 10 iterations to converge. I did not run PKN cases yet. Do you see any performance issues there?
> > 
> > 
> > In this PR, I slightly updated the MGR recipe for HF; otherwise, un-convergence issue will be faced for PKN cases.
> 
> I also saw enormous residuals when debugging the quasi-newton for the penny shaped case, do you have any idea which equations are associated with this big number? @jhuang2601 @CusiniM

I scanned through the code. The reason is the scaling factor in the residual evaluation in "SinglePhaseBaseKernels.hpp". The default option I think is Linf norm.  Each residual is normalized by either its cell total mass at old states or the default normalizer (1e-12), whichever is bigger. Since the newly generated fracture cell volume can be very small or zero, it may use 1e-12. If you switch to L2 norm for the flow solver evaluation, I did not see the large residual number anymore. @jhuang2601  @CusiniM

**Comment by CusiniM** [comment]
> > > > > > I tested penny-shaped and KGD poroelastic FP. Results look good to me.
> > > > > 
> > > > > 
> > > > > Thanks @Guotong-Ren for your check. Do you have any chance to run PKN case? How about the performance of non-linear solver?
> > > > 
> > > > 
> > > > I only see only one time step size cut for both viscosity and toughness dominated penny shape cases. The NL solver performs well I tight the convergence criterion to 1e-6/-7 and each time step solve takes about 10 iterations to converge. I did not run PKN cases yet. Do you see any performance issues there?
> > > 
> > > 
> > > In this PR, I slightly updated the MGR recipe for HF; otherwise, un-convergence issue will be faced for PKN cases.
> > 
> > 
> > I also saw enormous residuals when debugging the quasi-newton for the penny shaped case, do you have any idea which equations are associated with this big number? @jhuang2601 @CusiniM
> 
> I scanned through the code. The reason is the scaling factor in the residual evaluation in "SinglePhaseBaseKernels.hpp". The default option I think is Linf norm. Each residual is normalized by either its cell total mass at old states or the default normalizer (1e-12), whichever is bigger. Since the newly generated fracture cell volume can be very small or zero, it may use 1e-12. If you switch to L2 norm for the flow solver evaluation, I did not see the large residual number anymore. @jhuang2601 @CusiniM

Thanks for looking into this. It makes sense. Although, it's probably  not the best scaling choice then.

File: inputFiles/hydraulicFracturing/pennyShapedToughnessDominated_poroelastic_smoke.xml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+<Problem>
+  <Included>
+    <File
+      name="./pennyShapedToughnessDominated_poroelastic_base.xml"/>
+  </Included>
+  
+  <Solvers
+    gravityVector="{ 0.0, 0.0, -0.0 }">
+    <Hydrofracture
```
Comment (paveltomin) [inline inputFiles/hydraulicFracturing/pennyShapedToughnessDominated_poroelastic_smoke.xml:11]: @jhuang2601 @CusiniM looks like `isMatrixPoroelastic="1"` is missing here

File: inputFiles/hydraulicFracturing/pennyShapedViscosityDominated_poroelastic_smoke.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+<Problem>
+  <Included>
+    <File
+      name="./pennyShapedViscosityDominated_poroelastic_base.xml"/>
+  </Included>
+
+  <Solvers
+    gravityVector="{ 0.0, 0.0, -0.0 }">
+    <Hydrofracture
```
Comment (paveltomin) [inline inputFiles/hydraulicFracturing/pennyShapedViscosityDominated_poroelastic_smoke.xml:11]: and here
need to add `isMatrixPoroelastic="1"`

File: inputFiles/hydraulicFracturing/pknViscosityDominated_poroelastic_smoke.xml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+<Problem>
+  <Included>
+    <File
+      name="./pknViscosityDominated_poroelastic_base.xml"/>
+  </Included>
+
+  <Solvers
+    gravityVector="{ 0.0, 0.0, -0.0 }">
+    <Hydrofracture
```
Comment (paveltomin) [inline inputFiles/hydraulicFracturing/pknViscosityDominated_poroelastic_smoke.xml:11]: and also here
add `isMatrixPoroelastic="1"`

**Pull request merged**